File: android/src/main/kotlin/com/ovaom/flutter_libpd/pd_controller/pd_message_receiver/PdMessageReceiverImpl.kt
package com.ovaom.flutter_libpd.pd_controller.pd_message_receiver

import org.puredata.android.utils.PdUiDispatcher
import org.puredata.core.PdBase
import org.puredata.core.PdListener
import org.puredata.core.utils.PdDispatcher

class PdMessageReceiverImpl(private val dispatcher: PdDispatcher = PdUiDispatcher()): PdMessageReceiver, PdListener {
    private var callbacks = mutableMapOf<String, ListenerCallback>()

    init {
        PdBase.setReceiver(dispatcher)
    }

    override fun addListener(source: String, callback: ListenerCallback) {
        if (!callbacks.containsKey(source)) {
            dispatcher.addListener(source, this)
        }
        callbacks[source] = callback
    }

    override fun removeListener(source: String) {
        callbacks.remove(source)
        dispatcher.removeListener(source, this)
    }

    override fun removeAllListeners() {
        callbacks.keys.forEach { removeListener(it) }
    }

    override fun dispose() {
        removeAllListeners()
        PdBase.setReceiver(null)
    }

    override fun receiveBang(source: String) {
        callbacks[source]?.invoke(ReceivedEvent(source))
    }

    override fun receiveFloat(source: String, x: Float) {
        callbacks[source]?.invoke(ReceivedEvent(source, arguments = listOf(x)))
    }

    override fun receiveSymbol(source: String, symbol: String) {
        callbacks[source]?.invoke(ReceivedEvent(source, symbol))
    }

    override fun receiveList(source: String, vararg args: Any) {
        callbacks[source]?.invoke(ReceivedEvent(source, arguments = args.toList()))
    }

    override fun receiveMessage(source: String, symbol: String, vararg args: Any) {
        callbacks[source]?.invoke(ReceivedEvent(source, symbol, args.toList()))
    }
}

File: ios/Classes/PdController/PdMessageReceiver/PdMessageReceiverImpl.swift
//
//  PdMessageReceiverImpl.swift
//  flutter_libpd
//
//  Created by flav on 01/06/2022.
//

import Foundation
import libpd_ios

public class PdMessageReceiverImpl: NSObject, PdMessageReceiver {
    private let dispatcher: PdDispatcher
    private var callbacks:[String: ListenerCallback] = [:]
    
    public init(dispatcher: PdDispatcher = PdDispatcher()) {
        self.dispatcher = dispatcher
        PdBase.setDelegate(self.dispatcher)
    }
    
    public func addListener(source: String, callback: @escaping ListenerCallback) {
        if callbacks.doesNotContain(key: source) {
            dispatcher.add(self, forSource: source)
        }
        callbacks[source] = callback
    }
    
    public func removeListener(source: String) {
        callbacks.removeValue(forKey: source)
        dispatcher.remove(self, forSource: source)
    }
    
    public func removeAllListeners() {
        callbacks.keys.forEach { removeListener(source: $0) }
    }
    
    public func dispose() {
        removeAllListeners()
        PdBase.setDelegate(nil)
    }
}

extension PdMessageReceiverImpl: PdListener {
    
   public func receiveMessage(_ message: String, withArguments arguments: [Any], fromSource source: String) {
        callbacks[source]?(ReceivedEvent(source: source, symbol: message, arguments: arguments))
    }
    
    public func receive(_ received: Float, fromSource source: String) {
        callbacks[source]?(ReceivedEvent(source: source, arguments: [received]))
    }
    
    public func receiveSymbol(_ symbol: String, fromSource source: String) {
        callbacks[source]?(ReceivedEvent(source: source, symbol: symbol))
    }
    
    public func receiveBang(fromSource source: String) {
        callbacks[source]?(ReceivedEvent(source: source))
    }
    
    public func receiveList(_ list: [Any], fromSource source: String) {
        callbacks[source]?(ReceivedEvent(source: source, arguments: list))
    }
}

extension Dictionary where Key == String {
    func contains(key: Key) -> Bool {
        keys.contains(key)
    }
    
    func doesNotContain(key: Key) -> Bool {
        !contains(key: key)
    }
}
